### PR TITLE
Add render* functions to avoid show :bomb:s

### PR DESCRIPTION
**Pull request by olorin**
@charleso this kinda what you had in mind?

### CODE REVIEW

File: ambiata-p.cabal [hunk 1]
```diff
@@ -41,6 +42,7 @@ library
                        P.Function
                        P.Foldable
                        P.Debug
+                       P.Show
```
Comment (charleso) [inline ambiata-p.cabal]: Missing file?

**Comment by nhibberd** [comment]
@olorin missing a module maybe

**Comment by olorin** [comment]
Indeed, fixed.

File: ambiata-p.cabal [hunk 2]
```diff
@@ -18,6 +18,7 @@ library
                         , bifunctors                  >= 4.2        && <= 5.3
                         , deepseq                     >= 1.3        && <= 1.5
                         , semigroups                  >= 0.16       && < 0.19
+                        , text                        >= 1.1
```
Comment (charleso) [inline ambiata-p.cabal]: I suspect we've tried to avoid this in the past, but it would certainly be very convenient.

Comment (charleso) [inline ambiata-p.cabal]: In any case, should it have an upper bounds?

Comment (olorin) [inline ambiata-p.cabal]: Don't see why not, fixed.

**Comment by charleso** [comment]
I'm :100: to the idea. Not 100% sure on whether we should (for now) just expose the `String` versions of those functions. I suspect I'm being overly-cautious and it's fine, and if no-one has any objections go for it!

**Comment by olorin** [comment]
Unsure, does anyone know if there's a reason for not depending on `text` here? The `String` versions would be slightly less convenient, but still an improvement.

**Comment by thumphries** [comment]
would be nice to export `Text` too, if we're gonna do this. could also be `x-text` if the dep isn't ok

**Comment by olorin** [comment]
I dunno, should `Text` be exported unqualified? Might get a bit annoying e.g., when working on modules that only use `ByteString`s.

**Comment by thumphries** [comment]
i reckon every module in every project we have has

```haskell
import Data.Text (Text)
import qualified Data.Text as T
```

at the top

**Comment by thumphries** [comment]
> I dunno, should Text be exported unqualified?

just the type, to be clear - definitely not the module

**Comment by nhibberd** [comment]
I believe text is already a transitive dep. We could just re-export the type, not really safe to do anything else

**Comment by olorin** [comment]
Ah right, that makes more sense. `Text` type now re-exported.

**Comment by amosr** [comment]
I don't really understand the rationale [sic!!] behind this?

**Comment by olorin** [comment]
> I don't really understand the rationale [sic!!] behind this?

Basically because I keep fucking up and `show`ing stuff I didn't mean to, and the typechecker didn't complain because everything has a `Show` instance. Also because I'm sick of writing `T.pack . show` everywhere.

**Comment by thumphries** [comment]
https://github.com/search?q=user%3Aambiata+pack+.+show&ref=opensearch&type=Code

41 pages, almost all integrals

**Comment by nhibberd** [comment]
@amosr Should highlight that any use of show is probably doing the wrong thing, will make it easier to review the valid cases vs the invalid (invalid being using show on error types etc)

**Comment by tmcgilchrist** [comment]
Personally I don't find this as much of an issue. I try not to use show unless I really mean to.

**Comment by amosr** [comment]
fair enough then - I guess I don't come across this much in Icicle because we use a pretty printing library for most things

**Comment by olorin** [comment]
:+1: then? Any objections @ambiata/engineers?

**Comment by amosr** [comment]
no objections, and I believe exporting Text type alone is a boon :+1:

File: src/P.hs [hunk 3]
```diff
@@ -73,6 +74,7 @@ import           Data.Int as X
 import           Data.Ord as X
 import           Data.Tuple as X
 import           Data.Traversable as X
+import           Data.Text as X (Text)
```
Comment (nhibberd) [inline src/P.hs]: Is this going to conflict upstream with any current imports?
i.e.
```
import Data.Text (Text)
import P
```

Comment (olorin) [inline src/P.hs]: Yeah it will, re-exports aren't resolved wrt import conflicts I think. People will have to manually upgrade for things to break though.

**Pull request merged**